Make plotting with gradients more fun

NilsNemitz opened the pull request:
Inspired by #1712, I looked at what options we have for plotting with colors according to values.

It turns out that QPen lets us do that without problem, but currently it is necessary to manually define a gradient. As a user, it is also not obvious that the coordinates used in the definition of the gradient directly correspond to the plotted values.

This PR adds convenience methods to get a `QBrush` or `QPen` from a ColorMap object.
It extends the functionality of the already existing `getGradient()` method to handle repeating and mirrored mapping modes.

It also adds a `reverse()` method to ColorMap, which reverses an existing map, which is often convenient for imported maps.

An example is included based on the discussion of #1737, demonstrating that smoothly updating plots are possible even with fancy coloring options.

j9ac9k commented on examples/ColorGradientPlots.py:
```diff
@@ -0,0 +1,139 @@
+        )
+        self.timer = QtCore.QTimer(timerType=QtCore.Qt.PreciseTimer)
+        self.timer.timeout.connect(self.update)
+        timestamp = time.time()
```
Probably should use `time.perf_counter()` for float type, or `time.perf_counter_ns()`  for int type (in ns).  Also we have [ptime.py](https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/ptime.py) but I'm not sure it's actually relevant any more or not...

NilsNemitz replied:
(already updated)

j9ac9k commented:
I just ran the example, this is slick!

The one suggestion I would have is to add a title to each plot with a short description of how it is determined what color to apply.

NilsNemitz commented:
Dear @j9ac9k , I have added text labels to each plot to detail what they show. I have also switched to the performance counter, although that does not matter much as we are already averaging over a few frames to get a good number.

I also got Sphinx to work on my computer and did a full pass over the color map documentation.
Let me know what you think!

ixjlyons commented on doc/source/colormap.rst:
```diff
@@ -1,8 +1,92 @@
+
+To see all available color maps, please run the ColorMap demonstration available in the suite of :ref:`examples`.
+
+Examples
```
I'm a bit on the fence on this. I'm absolutely in favor of having small usage examples with visuals in the docs, but I think we should be careful about how to do it in a way that's scalable and won't lead to docs that are out of sync with the library. I've been working on an extension to [sphinx-gallery](https://sphinx-gallery.github.io/stable/index.html) which could be used to generate a gallery of the full example scripts. I've also dreamt of a sphinx extension to generate renderings of code blocks in the docs. I suppose we can always leave this in and replace it later if something like that ever gets going. Currently waiting on readthedocs to support installation of apt packages to install xvfb for rendering.

One thing we could do for now is use your scripts for generating these images and just list that code here via [`literalinclude`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-literalinclude) and appropriate line numbers. It might be worth converting the code to a more "script-like" structure in that case.

ixjlyons commented on pyqtgraph/colormap.py:
```diff
@@ -47,16 +55,25 @@ def listMaps(source=None):
+        `pyqtgraph/colors/maps/` folder for examples of the format.
+    source: str, optional
+        If omitted, a locally stored map is returned. Otherwise
+        - 'matplotlib' imports a map defined by Matplotlib.
```
If there's no blank line between bullet points, they don't get rendered correctly. I'd just remove the bullet points and write it as a single sentence.

NilsNemitz replied:
I added the missing blank lines, since I found this hard to read when it mixes the 'matplotlib' and 'colorcet' explanations in the same line.

ixjlyons commented on pyqtgraph/colormap.py:
```diff
@@ -47,16 +55,25 @@ def listMaps(source=None):
+    name: str
+        Name of color map. In addition to the included maps, this can also 
+        be a path to a file in the local folder. See the files in the
+        `pyqtgraph/colors/maps/` folder for examples of the format.
```
Double backticks for RST, annoying I know

NilsNemitz replied:
⇒ changed.

ixjlyons reviewed (commented):
Very nice demo. I've left some minor feedback mostly regarding RST formatting.

j9ac9k commented on pyqtgraph/colormap.py:
```diff
@@ -180,22 +197,23 @@ def getFromColorcet(name):
+    `cm[0.5]` returns a QColor corresponding to the center of ColorMap `cm`.
 
-    To provide user-defined color mappings, see :class:`GradientWidget <pyqtgraph.GradientWidget>`.
+    To provide interactively user-defined color mappings, see :class:`GradientWidget <pyqtgraph.GradientWidget>`.
```
a nice shorthand here would be 

```
:class:`~pyqtgraph.GradientWidget`
```

NilsNemitz replied:
⇒ changed.

j9ac9k commented on pyqtgraph/colormap.py:
```diff
@@ -330,25 +398,107 @@ def mapToFloat(self, data):
+        
+        Parameters
+        ----------
+        p1: QtCore.QpointF, default (0,0)
```
super nit-picky, I know but should be `QtCore.QPointF`

NilsNemitz replied:
⇒ changed.

NilsNemitz commented on doc/source/colormap.rst:
```diff
@@ -1,8 +1,92 @@
+
+Color maps can be user defined by assigning a number of *stops* over the range of 0 to 1. A color is given for each stop, and the in-between values are generated by interpolation.
+
+When map colors directly represent values, an improperly designed map can obscure detail over certain ranges of values, while creating false detail in others. PyQtGraph includes the perceptually uniform color maps provided by the `colorcet project` <https://colorcet.holoviz.org/>. Color maps can also be imported from the `colorcet` library or from `Matplotlib`, if either of these is installed.
```
⇒ changed.
Colorcet doesn't seem to quite agree on whether the name is capitalized or not :)

NilsNemitz commented on doc/source/colormap.rst:
```diff
@@ -1,8 +1,92 @@
+Color maps are commonly used to generate false color images, color scatter-plot points, 
+and illustrate the height of surface plots.
+
+PyQtGraph's ColorMap object can also be used a convenient source of colors from a 
```
⇒ changed.
I think a good approach is to have the first reference to `ColorMap` appear as the class reference, and the just use it as plain text through the rest of the page. Do we have other guidelines for that yet?

NilsNemitz commented on doc/source/colormap.rst:
```diff
@@ -1,8 +1,92 @@
-ColorMap
-========
+Color maps
```
⇒ changed.

NilsNemitz commented:
Dear @ixjlyons and @j9ac9k , thank you for the crash course in rst formatting. I am still making things up as I go along. I hope things work reasonably well now.

I have managed to directly pull the example code from the source files that generate the images. There is a convenient `dedent` parameter that takes out the extra spaces not needed in the examples.

I hope we'll get working support for the sphinx gallery at some point. When your library does graphics, it seems helpful to have images in the documentation!

Until it is a little clearer how this will work in practice, I'd suggest that we try to collect all the code that generates example images. This is in the images folder now, but that's only because I wasn't sure what the best location would be.

For the current sphinx format, a width of 420 pixels seems to work reasonably well for a half-width image, and that keeps the files small enough (17kB and 36 kB here) to just throw them into git for now; at least as long as we don't update them all the time :)

j9ac9k commented on pyqtgraph/colormap.py:
```diff
@@ -115,7 +135,7 @@ def _getFromFile(name):
 
 def getFromMatplotlib(name):
-    """ import colormap from matplotlib definition """
+    """ Generates a ColorMap object from a Matplotlib definition. Same as ``colormap.get(name, source='matplotlib')``. """
```
should this perhaps be 

```
:method:~`pyqtgraph.colormap.get`
```

that way we get links within the docs?

j9ac9k replied:
actually, scratch that, you're demonstrating an equivalent function, not looking to reference one...

j9ac9k commented on pyqtgraph/colormap.py:
```diff
@@ -221,29 +239,28 @@ class ColorMap(object):
-                            MIRROR maps colors to [0.0;-1.0] and [0.0;+1.0] identically
-                            DIVERGING maps colors to [-1.0;+1.0]
-        ===============     =======================================================================
+        __init__(pos, color, mapping=ColorMap.CLIP)
```
just out of curiosity, why is this line needed?  I see the output renders fine; but I'm just curious.

NilsNemitz replied:
It makes it so that `ColorMap.CLIP` is spelled out just like that. The automatic signature shows the numerical value, and `mapping=2` is not so helpful.

j9ac9k commented on pyqtgraph/colormap.py:
```diff
@@ -221,29 +239,28 @@ class ColorMap(object):
+            List of colors, interpreted via 
+            :func:`mkColor() <pyqtgraph.mkColor>`.
+        mapping: str or int, optional
+            Controls
```
get rid of newline?

NilsNemitz replied:
Probably. 
Sphinx has some kind of obscure method to determine line breaks in the parameter definitions, though.

With the bonus newline, this renders as 
mapping _(str or int, optional)_ - Controls [text continues]

Without the bonus newline, I _think_ this rendered as
mapping _(str or int, optional)_ -
Controls [text continues]

I find the hanging dashes ugly...
But I agree that it is probably not worth fighting sphinx over.

⇒ will remove

j9ac9k commented:
Should have merged this days ago, sorry for the wait @NilsNemitz This LGTM.  Thanks for jumping on this and taking the opportunity from the now-closed issue.
